Matcher in loader, take 4

cro opened the pull request:
### What does this PR do?

Creates a new directory `salt/matchers` and migrates the matcher functions into separate .py files in this directory.  These functions are loaded at minion startup, but can be reloaded with the new `saltutil.refresh_matchers` function.

### Tests written?

Matcher tests apply here.

### Commits signed with GPG?

Yes

cro committed:
Add matchers directory and move default set of matchers to separate .py

cro committed:
Add matchers directory and move default set of matchers to separate .py

cro committed:
Add matchers directory and move default set of matchers to separate .py

cro committed:
Add some more needed imports

cro committed:
Handle this as a classmethod amd be smarter about setting it up.

cro committed:
Typo

cro committed:
Clarify setup

cro committed:
Lint

cro committed:
Remove unused imports

cro committed:
Split matchers out into separate .py files.  Add refresh_matchers, sync_matchers.

cro committed:
Refactor matchers to all use similar function signatures.

cro committed:
Add docs for loadable matchers.

cro committed:
Add master-side cache matcher

cro committed:
Add compound pillar_exact_match

cro commented:
Previous incarnations:

#48624
#48784
#48785

cro committed:
Fix test.

gtmanfred commented on tests/integration/netapi/rest_tornado/test_app.py:
```diff
@@ -203,8 +203,8 @@ def test_multi_local_async_post(self):
         self.assertIn('jid', ret[0])
         self.assertIn('jid', ret[1])
-        self.assertEqual(ret[0]['minions'], sorted(['minion', 'sub_minion', 'localhost']))
-        self.assertEqual(ret[1]['minions'], sorted(['minion', 'sub_minion', 'localhost']))
```
Why are these getting removed?

This was added for testing the `enable_ssh_minions` feature added in 2018.3.

It does need the ssh daemon started in runtests.py with `python -m tests.runtests --ssh`, but those minions should still be returning from the tornado tests.

cro replied:
OK, this is what I needed to know, thanks.

gtmanfred reviewed (changes requested):
After talking to Nicole, those tests are not the flaky ones that she marked in 2017.7, so they should still be passing.

These should be returning localhost, because it should be matching the minions for the `enable_ssh_minions` targeting. in salt.utils.minions

https://github.com/saltstack/salt/blob/develop/salt/utils/minions.py#L715

isbm reviewed (approved):
@cro :boom: very cool! :boom:

gtmanfred commented:
@rallytime looks like these tests are failing on the head of develop too? can you confirm?

@cro revert the test changes, and I will fix them in a different PR, and approve this one.

gtmanfred commented:
Trying the tests inside of test-kitchen to see if I can verify them in our test setup.

cro committed:
Revert "Fix test."

This reverts commit 86b5c95c6cece645e69f098d3dd88220435a569b.

cro committed:
lint

cro committed:
Remove default.py

cro committed:
Lint

cachedout commented on doc/topics/matchers/index.rst:
```diff
@@ -0,0 +1,81 @@
+
+Matchers are modules that provide Salt's targeting abilities.  As of the
+Flourine release, matchers can be dynamically loaded.  Currently new matchers
+cannot be created, but existing matchers may have their functionality altered or
```
Is this just because the CLI would need to be modified as well? I feel like this needs more explanation.

cachedout reviewed (commented):
My big concern here is that these functions all take `self` and then get converted on-the-fly to class-based methods from the context of the minion.

I presume that this was to avoid changing the interface to the Matchers from the minion's perspective but I'm not sure this is the right decision. IMHO, we should modify the minion-side logic so that it just calls functions. This is the pattern in every other pluggable system in Salt and I am very hesitant about diverging from that.

rallytime commented:
@cro I have changed the base branch of this PR from develop to fluorine for inclusion in the Fluorine feature release.

Also, if you could make the changes requested above ASAP, that would be great so we can get this merged in. Thanks!

cro commented:
So if I understand correctly @cachedout, you would like to see this behave more like executors, which are loaded specifically with a call to `salt.loader.executors()`, and then referenced with the standard `__salt__['...']` notation?

cro committed:
Remove Matcher class and use the bare loader.

cro commented:
K, `d62b87d` there removes the Matcher class and goes instead with the bare loader.

cro commented:
Tests may not pass fully yet, I have a mock or two I have to figure out what to do with.

cro committed:
Add logging back into confirm_top.py, lint.

cachedout commented:
@cro Correct. I know that's a bit of a pain but the consistency will be a win in the long run. Thanks for doing this!

cachedout commented:
@cro Also a syntax error snuck in: https://jenkinsci.saltstack.com/job/pr-lint/job/PR-48809/13/warnings52Result/new/

cro committed:
Typo.

cro committed:
Typo.

cachedout commented:
@cro More fixing needed: https://jenkinsci.saltstack.com/job/pr-kitchen-centos7-py2/job/PR-48809/15/console

rallytime commented:
@cro Any update here? We need these tests fixed before we can get this in.

cro committed:
Remove `self`

cro committed:
Remove `self`
Fix vimfail in modules/match.py

cro committed:
Clarify doc

cro committed:
Lint

cro committed:
Lint

rallytime commented:
@cro There are some related test failures here. I am not sure about the tornado ones as sometimes those are flaky, but these failed everywhere in the same way. Has this been rebased recently? We've been trying to get the tests in a more stable place, so they might be related. The following list is definitely related though: 

- integration.minion.test_pillar.BasePillarTest.test_pillar_top_compound_match
- integration.shell.test_matcher.MatchTest.test_nodegroup
- integration.shell.test_matcher.MatchTest.test_salt_documentation_arguments_not_assumed
- unit.test_pillar.PillarTestCase.test_topfile_order

Can you please fix them?

cro committed:
Back out list matcher optimization, somehow breaks nodegroup matching.

rallytime commented:
@cro It appears now there is an error here. The minion won't start in the tests because of the following:
```
11:50:42        15:50:31,709 [salt.master                              :1840][ERROR   ] Error in function _pillar:
11:50:42        Traceback (most recent call last):
11:50:42          File "/tmp/kitchen/testing/salt/master.py", line 1833, in run_func
11:50:42            ret = getattr(self, func)(load)
11:50:42          File "/tmp/kitchen/testing/salt/master.py", line 1538, in _pillar
11:50:42            data = pillar.compile_pillar()
11:50:42          File "/tmp/kitchen/testing/salt/pillar/__init__.py", line 994, in compile_pillar
11:50:42            matches = self.top_matches(top)
11:50:42          File "/tmp/kitchen/testing/salt/pillar/__init__.py", line 655, in top_matches
11:50:42            self.opts.get('nodegroups', {}),
11:50:42          File "/tmp/kitchen/testing/salt/matchers/confirm_top.py", line 36, in confirm_top
11:50:42            return m(match)
11:50:42          File "/tmp/kitchen/testing/salt/matchers/compound_match.py", line 97, in match
11:50:42            six.text_type(matchers['{0}_match.match'.format(engine)](*engine_args, **engine_kwargs))
11:50:42        TypeError: match() takes exactly 2 arguments (1 given)
```

cro committed:
Backed out the change too far :-(

cro committed:
Pass __opts__ in modules/match.py since we no longer have self.

terminalmage committed:
Optimize list matcher by doing string membership checks,

This is faster than splitting into a list and then iterating over it.

terminalmage committed:
Optimize subnet membership check

By using a generator comprehension, we avoid pre-allocating an entire
list of objects.

terminalmage committed:
Optimize subdict matching

This reverses the order in which matches are attempted so that we start
with more deeply-nested matches first. The rationale for this is that it
is much more likely that when one uses `foo:bar:baz:hello:world`, they
are looking for a deeply-nested dictionary containing a key `hello`
which maps to `world`, than a key named `foo` matching the string
`bar:baz:hello:world`. The one case in which this change would be less
performant would be in the case where the value being checked is a
UNIX-style PATH (which uses colons as separators).

terminalmage committed:
Also support passing address as a tuple

*Just* in case something else that invokes this starts using tuples at
some point, it'll save us a traceback.

isbm reviewed (commented):
Seems too much test failures so far.

cro committed:
Try list matcher optimization again.

cro committed:
Imported wrong `six`.

rallytime commented:
@cro Getting closer, but there are still some tests that need to be fixed here:

https://jenkinsci.saltstack.com/job/pr-kitchen-centos7-py3/job/PR-48809/24/

All of those test failures are related to the changes in this PR.

terminalmage committed:
Make sequence optimization more efficient

terminalmage commented on salt/matchers/list_match.py:
```diff
@@ -3,16 +3,23 @@
-            or tgt.startswith(__opts__['id'] + ',') \
-            or tgt.endswith(',' + __opts__['id'])
+        if isinstance(tgt, collections.Sequence) and not isinstance(tgt, six.string_types):
+            result = bool(__opts__['id'] in tgt)
```
`__opts__['id'] in tgt` already returns a bool, so wrapping it in `bool()` is redundant.

Also, we can skip the type checking altogether by just relying on exception handling to narrow down what type the target is. I've opened https://github.com/cro/salt/pull/16 with one possible way we could do this.

cro committed:
Merge pull request #16 from terminalmage/matcher-optimizations

Make sequence optimization more efficient

terminalmage committed:
Add target to the warning log message

terminalmage committed:
Fix netapi tests

localhost should not appear in minion list unless there is an exchanged
key for it, which there is not

terminalmage committed:
Add a dunder init for matchers

terminalmage committed:
Fix pillar unit tests to work properly with loader-based matchers

terminalmage committed:
lint

terminalmage commented:
AWS is having some issues, but tests did pass on the one platform where they finished without SSH connection issues. This should be good to merge once we can get the test suite to run to completion.
